oci/layout/oci_transport.go

```diff
@@ -228,7 +247,75 @@ func (ref ociReference) NewImageDestination(ctx context.Context, sys *types.Syst
 
 // DeleteImage deletes the named image from the registry, if supported.
 func (ref ociReference) DeleteImage(ctx context.Context, sys *types.SystemContext) error {
-	return errors.New("Deleting images not implemented for oci: images")
+
+	// Get the manifest for the image
+	manifestDescriptor, err := ref.getManifestDescriptor()
+	if err != nil {
+		return err
+	}
+
+	manifest, err := ref.getManifest(manifestDescriptor)
+	if err != nil {
+		return err
+	}
+
+	// Get all the layers used by all other images
+	index, err := ref.getIndex()
+	if err != nil {
+		return err
+	}
+	layersUsedByOtherImages := make(map[digest.Digest]bool, len(index.Manifests))
+	for _, v := range index.Manifests {
+		if v.Digest != manifestDescriptor.Digest {
+			otherImageManifest, err := ref.getManifest(v)
+			if err != nil {
+				return err
+			}
+			for _, layer := range otherImageManifest.Layers {
```

Comment:
- The manifests themselves are also stored as blobs.
- `Config.Digest` also needs to be accounted for.
- IIRC it’s possible for an entry in the top-level index to be an `imgspecv1.Index` again — at least this code does that to one level of nesting, the spec seems to suggest an arbitrary depth.

With OCI artifacts storing ~arbitrary data, it seems quite plausible for an entry in `Layers` refer to the same blob as a config/manifest/index of another image/artifact.

Reply:
Spec says implementation should support this: https://github.com/opencontainers/image-spec/blob/main/image-index.md?plain=1#L44

It also means that the WIP you mentioned earlier about returning an index when calling `getManifestDescriptor` which would be impossible if the image is in a nested index, or that'll be the index in the index it was found, not necessarily `index.json`

My head starts to hurt :-)

Reply:
(My personal opinion is that the nested indexes provide too much flexibility and the semantics are unclear for consumers, so I’m not that much of a fan. But the feature exists.)

There’s a bit of nuance in here: if the repo contains a nested index, it’s not ideal, but also not a _disaster_ when this code does not support that in `ImageSource` / `getManifestDescriptor`; the code would just “safely” fail and users would have to use some other tool.

OTOH if the deletion code does not handle nested indices, that could lead to data loss, so that is a more urgent concern.

So one approach would be to recursively enumerate the used blobs; another, I think quite reasonable, possibility would be to just fail when a nested index is encountered, and to tell the user to use another tool — that’s analogous to the `ImageSource` failure mode, but still safe.

---

Note that the above applies to “deeply nested” OCI indexes. One level of nesting is supported, to represent a multi-arch image. See the outcome of

```sh
skopeo --insecure-policy copy --all docker://quay.io/libpod/alpine oci:$dest
```

So… I haven’t actually written the code, but it seems to me that a full recursion (or possibly a “queue of blobs+blob types to scan”) support might be the easiest way to implement all of this anyway, instead of somehow special-casing the “one level of nesting” case.

Reply:
You are right that handling this properly in the deletion has to be extra safe.
The good news is that I now have a way to cleanly create nested indexes for tests so this can be tested properly (which uses recursion so in for a penny, in for a pound).

Thanks for the skopeo example, I'll have a look at how this is handled, that will surely be instructive.

Reply:
I've been inspecting the result from the `skopeo` command and it has been very educating :-)

One thing I noticed, reading `getManifestDescriptor`, is that you can toast your registry if you do something like this:
```
skopeo copy docker://quay.io/libpod/alpine:3.2 oci:/tmp/oci-registry/alpine
skopeo copy docker://quay.io/libpod/alpine:3.10.2 oci:/tmp/oci-registry/alpine
```

The index.json file will then contain two manifests (image manifests or sub indexes if you used the --all) which don't have any annotations (so no `org.opencontainers.image.ref.name`).

At that point, that directory is toast it seems:
```
skopeo copy oci:/tmp/oci-registry/alpine:3.2 oci:/tmp/oci-registry/alpine-copy 
FATA[0000] initializing source oci:/tmp/oci-registry/alpine:3.2: no descriptor found for reference "3.2" 
skopeo copy oci:/tmp/oci-registry/alpine oci:/tmp/oci-registry/alpine-copy 
FATA[0000] initializing source oci:/tmp/oci-registry/alpine:: more than one image in oci, choose an image
```

Would this be worthy of being reported as a bug? 
Sorry for the digression, I'm still in the learning phase :-)

Reply:
> At that point, that directory is toast, it seems:

It’s… a valid OCI structure. And the WIP `oci:/tmp/oci-registry/alpine:@0` syntax will eventually allow c/image to consume it.

I agree it’s very awkward that c/image can produce a structure that it can’t read.

Reply:
Can you point me to that WIP with the index? It could be useful, for me, to have a  look.

Reply:
That’s #1381 / #1677 . Out of date for a while, I’m afraid.